lib/IDE/CompletionInstance.cpp

```diff
@@ -0,0 +1,360 @@
+//===--- CompletionInstance.cpp -------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/IDE/CompletionInstance.h"
+
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/DiagnosticEngine.h"
+#include "swift/AST/DiagnosticsFrontend.h"
+#include "swift/AST/Module.h"
+#include "swift/AST/SourceFile.h"
+#include "swift/Basic/LangOptions.h"
+#include "swift/Basic/SourceManager.h"
+#include "swift/Driver/FrontendUtil.h"
+#include "swift/Frontend/Frontend.h"
+#include "swift/Parse/PersistentParserState.h"
+#include "swift/Subsystems.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/ADT/Hashing.h"
+
+using namespace swift;
+using namespace ide;
+
+std::unique_ptr<llvm::MemoryBuffer>
+swift::ide::makeCodeCompletionMemoryBuffer(const llvm::MemoryBuffer *origBuf,
+                                           unsigned &Offset,
+                                           StringRef bufferIdentifier) {
+
+  auto origBuffSize = origBuf->getBufferSize();
+  if (Offset > origBuffSize)
+    Offset = origBuffSize;
+
+  auto newBuffer = llvm::WritableMemoryBuffer::getNewUninitMemBuffer(
+      origBuffSize + 1, bufferIdentifier);
+  auto *pos = origBuf->getBufferStart() + Offset;
+  auto *newPos =
+      std::copy(origBuf->getBufferStart(), pos, newBuffer->getBufferStart());
+  *newPos = '\0';
+  std::copy(pos, origBuf->getBufferEnd(), newPos + 1);
+
+  return std::unique_ptr<llvm::MemoryBuffer>(newBuffer.release());
+}
+
+namespace {
+/// Returns index number of \p D in \p Decls . If it's not found, returns ~0.
+template <typename Range>
+unsigned findIndexInRange(Decl *D, const Range &Decls) {
+  unsigned N = 0;
```

Comment:
Is `Range` always random-access? If so, maybe simplify to

```c++
auto I = std::find(Decls.begin(), Decls.end(), D);
return I == Decls.end() ? ~0u : std::distance(Decls.begin(), I);
```

Reply:
`IterableDeclContext::getMembers()` returns `DeclRange` (`iterator_range<DeclIterator>`) which is unfortunately non-random-access.